### ISSUE TITLE
Training speed and GPU utility

### ISSUE THREAD

**Issue by crazylyf** [thtrieu/darkflow#306]
Hi, I'm training tiny-yolov2 on one TITAN Xp with `--gpu 0.9` and `--batch 64`, and found the training speed is very slow, say 8500 steps / 15 hrs. CPU utility keeps ~400%, but GPU utility is mostly 0% and only occasionally ~100%. Is this normal? What's the most time-consuming step?

**Comment by melaanya** [comment]
Mayve I'm wrong but it seems to me that I have the similar issue: option --gpu 0.5, but according to nvidia-smi it's never used. I don't know whether it is a problem of my wrong launching or something else.

**Comment by jubjamie** [comment]
Do you have the gpu version of tensorflow installed in the environment you are training yolo in?

**Comment by crazylyf** [comment]
@jubjamie  Yes, it is gpu version.

**Comment by melaanya** [comment]
@jubjamie Yes. I have gpu version too. I use Docker (official docker image tensorflow/tensorflow:latest-gpu) and videocard GeForce GTX 1080 8 Gb, Driver Version: 375.39.
Now I made it to fill all gpu memory but still the speed of training is very low - approximately 6000 iterations in 36 hours. In original (pjreddie) version of yolo it's possible to get 82000 iterations in 2 days. 
Am I doing everything right and it s correct speed of training on gpu?

**Comment by zhyj3038** [comment]
same problem

**Comment by jubjamie** [comment]
Hmm this is very odd. Does your GPU memory usage remain full? There will be some times when the CPU has to take over (e.g. writing to file etc.). I can't run any timing tests at the moment but 6000 iterations in 36 hours seems rather slow. I think I changed by batch size to 8 as I was getting memory issues. But I was running about 4 steps every 5 seconds.

There are a number of other process that happen during training including reading the data and annotations from disk. These little things may add up. I can't really help at this time but perhaps @abagshaw  or @thtrieu will have some kind of timing performance information to hand.

**Comment by crazylyf** [comment]
The most time-consuming steps are the preprocessing steps, among which the function `imcv2_recolor` defined in `utils/im_transform.py` is very slow!

**Comment by crazylyf** [comment]
After changing the `imcv2_recolor` function to a simpler alternative, the training can be speed up by 2-3x

**Comment by melaanya** [comment]
@crazylyf Can you tell, please, what alternative for imcv2_recolor you choosed?

**Comment by crazylyf** [comment]
@melaanya  I just rescale each each channel independently.

**Comment by jubjamie** [comment]
@crazylyf any chance of a code example or link to a library that you've used? Would be interesting to test. Although I imagine that @thtrieu had a reason for using the currently implemented method?

**Comment by crazylyf** [comment]
@jubjamie Sorry that I could not paste the code directly before being approved. The idea is quite simple: instead of original exponential scaling for each channel, I rescaled each channel with a random value.

**Comment by crazylyf** [comment]
Since the most time-consuming step is `np.power` function, which takes ~70ms for an (640, 480, 3)-sized ndarray, and tensorflow's `pow` function is much faster (>10x), I think we can also use `tf.pow` to speed up.

**Comment by jubjamie** [comment]
Does this only speed up training or also inference? I'm still unsure what you mean by rescaling each channel linearly?

**Comment by crazylyf** [comment]
@jubjamie It speed up only in training step, as there is no color perturbation for inference.  As to the channel rescaling, for each image channel, I scale all pixel values with a random factor, say 0.98.  you can customize the range of scale factors, and be careful of values larger than 255.

**Comment by jubjamie** [comment]
Oh is it for distorting the images during training?

**Comment by crazylyf** [comment]
@jubjamie Yes. Sorry for late reply.

**Comment by ryanferg** [comment]
Just wanted to let y'all know that this  #https://github.com/thtrieu/darkflow/pull/404 pull request fixes the slow numpy power function in imcv2_recolor. It led to a 3x speed-up for my training.
